src/stravalib/protocol.py

```diff
@@ -89,6 +102,190 @@ def __init__(
             lambda _request_params, _method: None
         )
 
+    def _check_credentials(self) -> tuple[int, str] | None:
```

Comment:
This is a new helper that checks credentials. This makes it easier to test and hopefully maintain and avoids adding a lot of logic to the refresh steps.

Reply:
Great refactoring. We can simplify this by always returning a tuple of potentially nullables (so the return type is `tuple[int | None, str | None]`) and storing the result as instance variables of `APIV3` for later reuse. See the other comments.